### PR TITLE
Optimize polyfit

### LINKED ISSUE #5629: Polyfit performance on large datasets - Suboptimal dask task graph
**What happened**:

I am trying to calculate a linear trend over a large climate model simulation. I use rechunker to chunk the data along the horizontal dimensions and make sure that the time dimension (along which I want to calculate the fit) is not chunked. 
In my realistic example, this blows up the memory of my workers. 

**What you expected to happen**:
I expected this to work very smoothly because it should be embarassingly parallel (no information of sourrounding chunks is needed and the time dimension is complete in each chunk).

**Minimal Complete Verifiable Example**:
I think this minimal example shows that the task graph created is not ideal

```python
import xarray as xr
import dask.array as dsa
da = xr.DataArray(dsa.random.random((4,6, 100), chunks=(1,2,100)), dims=['x','y', 'time'])
da
```

image

> Note that the original data array has 12 chunks

When I apply polyfit I get this
```python
poly = da.polyfit('time', 1)
poly
```

image

Now the number of chunks has decreased to 4? I am not sure why, but this indicates to me that my problem might be related to #4554.

When I look at the task graph it seems that this explains why for very large dataset the computation blows up:
image

As I said before I would expect this calculation to be fully parallelizable, but there seems to be some aggregation/rechunking steps in the bottom layer. How 'bad' these get (e.g. how many input chunks get lumped together in the `rechunk/merge` step, depends on the chunking structure of the horizontal axes. 

cc @TomNicholas

**Anything else we need to know?**:

**Environment**:

<details><summary>Output of <tt>xr.show_versions()</tt></summary>

INSTALLED VERSIONS
------------------
commit: None
python: 3.8.8 | packaged by conda-forge | (default, Feb 20 2021, 16:22:27) 
[GCC 9.3.0]
python-bits: 64
OS: Linux
OS-release: 5.4.89+
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: C.UTF-8
LANG: C.UTF-8
LOCALE: en_US.UTF-8
libhdf5: 1.10.6
libnetcdf: 4.7.4

xarray: 0.17.0
pandas: 1.2.4
numpy: 1.20.2
scipy: 1.6.2
netCDF4: 1.5.6
pydap: installed
h5netcdf: 0.11.0
h5py: 3.2.1
Nio: None
zarr: 2.7.1
cftime: 1.4.1
nc_time_axis: 1.2.0
PseudoNetCDF: None
rasterio: 1.2.2
cfgrib: 0.9.9.0
iris: None
bottleneck: 1.3.2
dask: 2021.04.1
distributed: 2021.04.1
matplotlib: 3.4.1
cartopy: 0.19.0
seaborn: None
numbagg: None
pint: 0.17
setuptools: 49.6.0.post20210108
pip: 20.3.4
conda: None
pytest: None
IPython: 7.22.0
sphinx: None

</details>

### PR DESCRIPTION
**Pull request by dcherian**
Closes #5629

1. Use Variable instead of DataArray
2. Use `reshape_blockwise` when possible following https://github.com/pydata/xarray/issues/5629#issuecomment-960133879

- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

cc @phofl

### CODE REVIEW

File: xarray/tests/test_dataset.py [hunk 1]
```diff
@@ -6698,7 +6698,7 @@ def test_polyfit_coord(self) -> None:
 
         out = ds.polyfit("numbers", 2, full=False)
         assert "var3_polyfit_coefficients" in out
-        assert "dim1" in out
+        assert "dim1" in out.dims
```
Comment (dcherian) [inline xarray/tests/test_dataset.py:6701]: this test was adding `dim1` as a *new* coordinate var (it is just a dimension name in `ds`). I *think* that's wrong.

cc @aulemahal

Comment (aulemahal) [reply xarray/tests/test_dataset.py:6701]: Indeed, that seems better!

File: doc/whats-new.rst [hunk 2]
```diff
@@ -29,6 +29,8 @@ New Features
 - Support lazy grouping by dask arrays, and allow specifying ordered groups with ``UniqueGrouper(labels=["a", "b", "c"])``
   (:issue:`2852`, :issue:`757`).
   By `Deepak Cherian <https://github.com/dcherian>`_.
+- Optimize :py:meth:`DataArray.polyfit` and :py:meth:`Dataset.polyfit` with dask, when possible.
```
Comment (dcherian) [inline doc/whats-new.rst:32]:
```suggestion
- Optimize :py:meth:`DataArray.polyfit` and :py:meth:`Dataset.polyfit` with dask, when used with 
   arrays with more than 2D dimensions.
```

**Comment by dcherian** [comment]
@aulemahal can you check that xclim's test suite passes with this please

**Comment by phofl** [comment]
Nice!

File: doc/whats-new.rst [hunk 3]
```diff
@@ -29,6 +29,9 @@ New Features
 - Support lazy grouping by dask arrays, and allow specifying ordered groups with ``UniqueGrouper(labels=["a", "b", "c"])``
   (:issue:`2852`, :issue:`757`).
   By `Deepak Cherian <https://github.com/dcherian>`_.
+- Optimize :py:meth:`DataArray.polyfit` and :py:meth:`Dataset.polyfit` with dask, when used with
+  arrays with more than 2D dimensions.
```
Comment (max-sixty) [inline doc/whats-new.rst:33]:
```suggestion
  arrays with more than two dimensions.
```

**Comment by aulemahal** [comment]
@dcherian I can confirm the test suite of xclim passes successfully when tested against this branch!

**Pull request merged**